### PR TITLE
[Serve] Introducing GPU sampler for CUDA

### PR DESCRIPTION
**Pull request by MasterJH5574**
This PR introduces the GPU sampler for CUDA only. The GPU sampler makes use of the GPU sampling ops introduced in apache/tvm#16575.

We will follow up to benchmark the performance of the GPU sampler over CPU sampler.

### CODE REVIEW

**Comment by MasterJH5574** [comment]
Marked as draft since it depends on apache/tvm#16703.

**Comment by MasterJH5574** [comment]
Special thanks for @yongwww for the great contribution on the sampling operators, which work smooooothly!

File: cpp/serve/function_table.cc [hunk 1]
```diff
@@ -221,6 +221,12 @@ void FunctionTable::_InitFunctions() {
   this->kv_cache_popn_func_ = get_global_func("vm.builtin.paged_attention_kv_cache_popn");
   this->kv_cache_get_num_available_pages_func_ =
       get_global_func("vm.builtin.paged_attention_kv_cache_get_num_available_pages");
+  if (local_gpu_device.device_type == DLDeviceType::kDLCUDA) {
```
Comment (tqchen) [inline cpp/serve/function_table.cc:224]: add function name for readability GPUSamplerSupported(local_gpu_device)

File: cpp/serve/model.cc [hunk 2]
```diff
@@ -284,6 +284,15 @@ class ModelImpl : public ModelObj {
                           std::move(trace_recorder));
   }
 
+  Sampler CreateSampler(int max_num_sample, Optional<EventTraceRecorder> trace_recorder) {
+    if (device_.device_type == DLDeviceType::kDLCUDA) {
```
Comment (tqchen) [inline cpp/serve/model.cc:288]: static bool Sampler::SupportGPUSampler(device_)

File: python/mlc_llm/compiler_pass/attach_to_ir_module.py [hunk 3]
```diff
@@ -3,7 +3,8 @@
 from typing import Any, Dict
 
 import tvm
-from tvm import IRModule, relax, tir
+from tvm import IRModule, relax, te, tir
+from tvm.relax.frontend import nn
```
Comment (tqchen) [inline python/mlc_llm/compiler_pass/attach_to_ir_module.py:7]: readability: I think it helps to rename to 

- attach_sampler.py
- attach_logit_processor.py
- attach_embedding_output_allocator.py
- attach_support_info.py

Comment (MasterJH5574) [reply python/mlc_llm/compiler_pass/attach_to_ir_module.py]: Agreed! Now split the file into multiple ones.

**Pull request merged**